feat: course outline section list

navinkarkera opened the pull request:
Continues work from https://github.com/openedx/frontend-app-course-authoring/pull/599

This PR adds:
* Section list
* Empty section list
* Publish modal
* Highlights modal
* Delete modal
* Edit section title
* Duplicate section option
* Fully functional collapse/expand option

<details><summary>Screenshots</summary>

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/10894099/935c12e3-8619-44b6-84df-959e8f49cc94)


![image](https://github.com/open-craft/frontend-app-course-authoring/assets/10894099/ae3e8ece-a427-40ac-abff-7fdabe10d242)

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/10894099/c1082e84-4527-4514-80b9-62db78f3dbe4)

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/10894099/bf23ebd4-bf29-46d2-b334-99d575c0f3d5)

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/10894099/e71a3465-11f7-4186-89df-0c2d13c502e4)

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/10894099/240cf781-3468-4c3b-8cf9-08eb2b862a6f)


</details>

**Dependents on**: 
* https://github.com/openedx/edx-platform/pull/33667
* https://github.com/openedx/frontend-app-course-authoring/pull/694

**Testing instructions**:

* Setup this repo as described in the readme.
* Checkout branch from https://github.com/openedx/edx-platform/pull/33667 and run `make cms-up` in devstack.
* Open any course outline by using this url format: http://localhost:2001/course/course-v1:edX+DemoX+Demo_Course/

vladislavkeblysh committed:
feat: course outline - sections list

* feat: [2u-259] add components

* feat: [2u-259] fix sidebar

* feat: [2u-259] add tests, fix links

* feat: [2u-259] fix messages

* feat: [2u-159] fix reducer and sidebar

* feat: [2u-259] fix reducer

* feat: [2u-259] remove warning from selectors

* feat: [2u-259] remove indents

---------

Co-authored-by: Vladislav Keblysh <vladislavkeblysh@Vladislavs-MacBook-Pro.local>

feat: Course Outline  - Sections list (#59)

* feat: [2u-336] add tests

* feat: [2u-271] fix button

* feat: [2u-336] add component, refactor header

* fix: [2u-342] fix translates and indents

* fix: [2u-342] fix constants and expand block

* feat: [2u-336] remove new section from menu

---------

Co-authored-by: Vladislav Keblysh <vladislavkeblysh@Vladislavs-MacBook-Pro.local>

feat: Course outline - Content empty (#72)

* feat: [2u-324] add component

* feat: [2u-324] add translates

* feat: [2u-324] update tests

* feat: [2u-324] update branch

* fix: [2u-324] fixed empty handler

feat: Course outline - Section Publish (#61)

* feat: [2u-354] add publish modal, api and update tests

* feat: [2u-354] refactor modal

* fix: [2u-354] removed comments

* fix: [2u-354] fix indents

* fix: [2u-354] removed translates duplicates

* fix: [2u-354] rename handlers

feat: Course outline - Update section card (#71)

* feat: [2u-615] update section card

* fix: [2u-615] fix handler names

* fix: [2u-615] fix indents

* fix: [2u-615] add empty handler

* fix: [2u-615] fix data test id name

* fix: [2u-615] fix styles

fix: [2u-696] add saving processing for higlights and enable highlights (#78)

feat: Course outline - Section Edit (#70)

* feat: [2u-336] add tests

* feat: [2u-271] fix button

* feat: [2u-336] add component, refactor header

* feat: [2u-342] add modal

* fix: [2u-342] fix translates and indents

* feat: [2u-342] add modal

* feat: [2u-342] add api

* feat: [2u-342] add tests and translates

* feat: [2u-342] fix indents

* fix: [2u-342] fix indents, variant and utils

* feat: [2u-342] fixed slice, thunks, hooks

* feat: [2u-354] add publish modal, api and update tests

* feat: [2u-615] update section card

* feat: [2u-348] add api, handlers, tests

* feat: [2u-348] add description for api

* fix: [2u-348] fix useEscapeClick

* fix: [2u-348] remove useEffect from CardHeader

* fix: [2u-348] fixed handlers and tests

* fix: [2u-348] fixed handlers and tests

---------

Co-authored-by: Vladislav Keblysh <vladislavkeblysh@Vladislavs-MacBook-Pro.local>

feat: Course outline - Section Delete (#74)

* feat: [2u-336] add tests

* feat: [2u-271] fix button

* feat: [2u-336] add component, refactor header

* feat: [2u-342] add modal

* fix: [2u-342] fix translates and indents

* feat: [2u-342] add modal

* feat: [2u-342] add api

* feat: [2u-342] add tests and translates

* feat: [2u-342] fix indents

* fix: [2u-342] fix indents, variant and utils

* feat: [2u-342] fixed slice, thunks, hooks

* feat: [2u-354] add publish modal, api and update tests

* feat: [2u-615] update section card

* feat: [2u-348] add api, handlers, tests

* feat: [2u-510] add delete api, add delete modal

* fix: [2u-510] fixed tests

---------

Co-authored-by: Vladislav Keblysh <vladislavkeblysh@Vladislavs-MacBook-Pro.local>

feat: Course outline - Section duplicate (#88)

* feat: [2u-336] add tests

* feat: [2u-271] fix button

* feat: [2u-336] add component, refactor header

* feat: [2u-342] add modal

* fix: [2u-342] fix translates and indents

* feat: [2u-342] add modal

* feat: [2u-342] add api

* feat: [2u-342] add tests and translates

* feat: [2u-342] fix indents

* fix: [2u-342] fix indents, variant and utils

* feat: [2u-342] fixed slice, thunks, hooks

* feat: [2u-354] add publish modal, api and update tests

* feat: [2u-615] update section card

* feat: [2u-348] add api, handlers, tests

* feat: [2u-510] add delete api, add delete modal

* feat: [2u-360] add api

* feat: [2u-360] add slice

* feat: [2u-360] add tests

* fix: [2u-360] fixed tests

---------

Co-authored-by: Vladislav Keblysh <vladislavkeblysh@Vladislavs-MacBook-Pro.local>

fix: Course outline - Highlights links (#89)

* fix: fixed doc urls

* fix: fixed components

feat: Course outline - Collapse all sections (#75)

* feat: added collapse all section logic

* fix: fixed tests

fix: final revision commits

fix: increase code coverage on the page

navinkarkera committed:
feat: section list and new section button

Also refactor api and hooks

fix: publish button behaviour and card header tests

fix: warning in highlights and publish modal test

fix: courseoutline tests

test: add test for new section functionality

fix(lint): lint issues

refactor: remove unnecessary css in CardHeader

refactor: rename emptyPlaceholder test file

refactor: replace ternary operator with 'and' condition

refactor: add black color to expand/collapse button

refactor: display only changed subsection and units in publish modal

Cup0fCoffee commented on src/course-outline/empty-placeholder/EmptyContent.test.jsx:
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
```
File should be named `EmptyPlaceholder.test.jsx` to match the component name.

Cup0fCoffee commented on src/course-outline/header-navigations/HeaderNavigation.test.jsx:
```diff
@@ -1,5 +1,6 @@
 import React from 'react';
```
Reminder that the file name has a typo. https://github.com/openedx/frontend-app-course-authoring/pull/694#discussion_r1405364001

Cup0fCoffee commented on src/course-outline/card-header/CardHeader.jsx:
```diff
@@ -0,0 +1,167 @@
+            data-testid="section-card-header__expanded-btn"
+            className={classNames('section-card-header__expanded-btn', {
+              collapsed: !isExpanded,
+            })}
```
Same as [here](https://github.com/openedx/frontend-app-course-authoring/pull/694#discussion_r1405364868), i.e. use `ArrowDownIcon` and `ArrowUpIcon` instead of css classes and styles to change orientation of the icon.

Cup0fCoffee commented on src/course-outline/card-header/CardHeader.scss:
```diff
@@ -0,0 +1,65 @@
+    & svg {
+      width: 1.5rem;
+      height: 1.5rem;
+    }
```
I don't think these are needed. The first one doesn't seem to do anything, because the button doesn't have `::before`, and the second one is not needed, because the child svg will take the same size as the parent container, which seems to be the same, i.e. `1.5x1.5 rem`.

Cup0fCoffee commented on src/course-outline/card-header/CardHeader.scss:
```diff
@@ -0,0 +1,65 @@
+
+    & span:first-child {
+      color: $black;
+    }
```
The color can be set on the parent button `.section-card-header__expanded-btn` the same way. I don't think this selector is necessary, and complicates overriding it using custom styles.

navinkarkera replied:
Removed, tertiary btn already has dark color.

Cup0fCoffee replied:
IIRC, it's not exactly black like the text, so it's understandable why the style was added, but I just don't think this selector is necessary, because the `span` can just inherit color from the parent button, and thus the styles can be simplified. I don't mind if you don't want to add it.

navinkarkera replied:
Oh, did not notice that the text color is black, added `color: $black` to the parent element.

Cup0fCoffee commented on src/course-outline/card-header/CardHeader.scss:
```diff
@@ -0,0 +1,65 @@
+  .section-card-header__menu {
+    display: flex;
+    align-items: center;
+  }
```
I don't think these are necessary, since the element is a button with only one child - span/icon, and should be already centered.

Cup0fCoffee commented on src/course-outline/publish-modal/PublishModal.jsx:
```diff
@@ -0,0 +1,77 @@
+      </ModalDialog.Header>
+      <ModalDialog.Body>
+        <p className="small">{intl.formatMessage(messages.description)}</p>
+        {subSections.length ? subSections.map((subSection) => {
+          const units = subSection.childInfo.children;
+
+          return units.length ? (
+            <React.Fragment key={subSection.id}>
+              <span className="small text-gray-400">{subSection.displayName}</span>
+              {units.map((unit) => (
+                <div
+                  key={unit.id}
+                  className="small border border-light-400 p-2 publish-modal__subsection"
+                >
+                  {unit.displayName}
+                </div>
+              ))}
+            </React.Fragment>
+          ) : null;
+        }) : null}
```
```suggestion
        {subSections.map((subSection) => {
          const units = subSection.childInfo.children;

          return units.length ? (
            <React.Fragment key={subSection.id}>
              <span className="small text-gray-400">{subSection.displayName}</span>
              {units.map((unit) => (
                <div
                  key={unit.id}
                  className="small border border-light-400 p-2 publish-modal__subsection"
                >
                  {unit.displayName}
                </div>
              ))}
            </React.Fragment>
          ) : null;
        })}
```

navinkarkera replied:
Updated this part such that only subsections/units which have changes are displayed.

Cup0fCoffee commented on src/course-outline/header-navigations/HeaderNavigations.jsx:
```diff
@@ -55,16 +56,18 @@ const HeaderNavigations = ({
-          ? intl.formatMessage(messages.collapseAllButton)
-          : intl.formatMessage(messages.expandAllButton)}
-      </Button>
+      {hasSections ? (
+        <Button
+          variant="outline-primary"
+          className={isSectionsExpanded ? 'expand-button-active' : ''}
+          iconBefore={ArrowDownIcon}
+          onClick={handleExpandAll}
+        >
+          {isSectionsExpanded
+            ? intl.formatMessage(messages.collapseAllButton)
+            : intl.formatMessage(messages.expandAllButton)}
+        </Button>
+      ) : null}
```
```suggestion
      {hasSections && (
        <Button
          variant="outline-primary"
          className={isSectionsExpanded ? 'expand-button-active' : ''}
          iconBefore={ArrowDownIcon}
          onClick={handleExpandAll}
        >
          {isSectionsExpanded
            ? intl.formatMessage(messages.collapseAllButton)
            : intl.formatMessage(messages.expandAllButton)}
        </Button>
      )}
```

Cup0fCoffee commented:
:+1: LGTM

- [x] I tested this
- [x] I read through the code
- [-] I checked for accessibility issues
- [-] Includes documentation

navinkarkera committed:
feat: course outline page (#694)

* feat: Course outline Top level page (#36)

* feat: [2u-259] add components

* feat: [2u-259] fix sidebar

* feat: [2u-259] add tests, fix links

* feat: [2u-259] fix messages

* feat: [2u-159] fix reducer and sidebar

* feat: [2u-259] fix reducer

* feat: [2u-259] remove warning from selectors

* feat: [2u-259] remove indents

---------

Co-authored-by: Vladislav Keblysh <vladislavkeblysh@Vladislavs-MacBook-Pro.local>

feat: Course outline Status Bar (#50)

* feat: [2u-259] add components

* feat: [2u-259] fix sidebar

* feat: [2u-259] add tests, fix links

* feat: [2u-259] fix messages

* feat: [2u-159] fix reducer and sidebar

* feat: add checklist

* feat: [2u-259] fix reducer

* feat: [2u-259] remove warning from selectors

* feat: [2u-259] remove indents

* feat: [2u-259] add api, enable modal

* feat: [2u-259] add tests

* feat: [2u-259] add translates

* feat: [2u-271] fix transalates

* feat: [2u-281] fix isQuery pending, utils, hooks

* feat: [2u-281] fix useScrollToHashElement

* feat: [2u-271] fix imports

---------

Co-authored-by: Vladislav Keblysh <vladislavkeblysh@Vladislavs-MacBook-Pro.local>

feat: Course Outline Reindex (#55)

* feat: [2u-277] add alerts

* feat: [2u-277] add translates

* feat: [2u-277] fix tests

* fix: [2u-277] fix slice and hook

---------

Co-authored-by: Vladislav Keblysh <vladislavkeblysh@Vladislavs-MacBook-Pro.local>

fix: Course outline tests (#56)

* fix: fixed course outline status bar tests

* fix: fixed course outline status bar tests

* fix: fixed course outline enable highlights modal tests

* fix: enable modal tests

fix: increase code coverage on the page

* refactor: improve course outline page

feat: lms live link

chore: update outline link

fix: course outline link

refactor: remove unnecessary css and rename test file

refactor: remove unnecessary css from outlineSidebar

test: make use of message variable instead of hardcoded text

refactor: remove unnecessary h5 class

test: use test id for detecting component

refactor: update course outline url and some default messages

---------

Co-authored-by: vladislavkeblysh <138868841+vladislavkeblysh@users.noreply.github.com>
